plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -130,6 +232,152 @@ class bp_connection_manager {
+            if (peer_info && peer_info->key) {
+               constexpr bool check_canonical = false;
+               public_key_type pk(peer.sig, peer.digest(), check_canonical);
+               if (pk != *peer_info->key) {
+                  fc_dlog(self()->get_logger(), "Recovered peer key did not match on-chain ${p}, recovered: ${pk} != expected: ${k}",
+                          ("p", peer.producer_name)("pk", pk)("k", *peer_info->key));
+                  return false;
+               }
+            } else { // unknown key
+               // ok, might have just been deleted or dropped out of top ranking
+               fc_dlog(self()->get_logger(), "Failed to find peer key ${p}", ("p", peer.producer_name));
+               return false;
+            }
+         } catch (fc::exception& e) {
+            fc_dlog(self()->get_logger(), "Exception recovering peer key ${p}, error: ${e}", ("p", peer.producer_name)("e", e.to_detail_string()));
+            fatal_error = true;
+            return false; // invalid key
+         }
+         return true;
+      };
+
+      fc::lock_guard g(gossip_bps.mtx);
+      auto& sig_idx = gossip_bps.index.get<by_sig>();
+      for (auto i = msg.peers.begin(); i != msg.peers.end() && !fatal_error;) {
+         const auto& peer = *i;
+         bool have_sig = sig_idx.contains(peer.sig); // we already have it, already verified
+         if (!have_sig && !is_peer_key_valid(peer)) {
+            // peer key may have changed or been removed on-chain, do not consider that a fatal error, just remove it
+            i = msg.peers.erase(i);
+         } else {
+            ++i;
+         }
+      }
+
+      if (fatal_error)
+         return false;
+      return !msg.peers.empty();
+   }
+
+   // thread-safe
+   bool update_gossip_bps(const gossip_bp_peers_message& msg) {
+      // providing us with full set
+      fc::lock_guard g(gossip_bps.mtx);
+      auto& idx = gossip_bps.index.get<by_producer>();
+      bool diff = false;
+      for (const auto& peer : msg.peers) {
+         if (auto i = idx.find(boost::make_tuple(peer.producer_name, boost::cref(peer.server_address))); i != idx.end()) {
+            if (i->sig != peer.sig) { // signature has changed, producer_name and server_address has not changed
+               gossip_bps.index.modify(i, [&peer](auto& m) {
+                  m.sig = peer.sig; // update the signature, producer_name and server_address has not changed
+               });
+               diff = true;
+            }
+         } else {
+            if (idx.count(peer.producer_name) >= max_bp_peers_per_producer) {
+               // only allow max_bp_peers_per_producer, choose one to remove
+               gossip_bps.index.erase(idx.find(peer.producer_name));
+            }
+            gossip_bps.index.insert(peer);
+            diff = true;
+         }
+      }
+      return diff;
+   }
+
+   flat_set<std::string> find_gossip_bp_addresses(const peer_name_set_t& accounts, const char* desc) const {
```

Comment:
Is there a reason to use a `flat_set` as a return value, as opposed to a `std::vector`?